ts/third_party/cloud-debug-nodejs/src/agent/io/scanner.ts

```diff
@@ -0,0 +1,264 @@
+/**
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as crypto from 'crypto';
+
+import * as events from 'events';
+// TODO: Make this more precise.
+const findit: (dir: string) => events.EventEmitter = require('findit2');
+
+import * as fs from 'fs';
+import * as _ from 'lodash';
+import * as path from 'path';
+
+// TODO: Make this more precise.
+const split: () => fs.WriteStream = require('split');
+
+export interface FileStats {
+  // TODO: Verify that this member should actually be optional.
+  hash?: string;
+  lines: number;
+}
+
+// TODO: Update the code so that `undefined  is not a possible property value
+export interface ScanStats {
+  [filename: string]: FileStats|undefined;
+}
+
+export interface ScanResults {
+  errors(): Map<string, Error>;
+  all(): ScanStats;
+  selectStats(regex: RegExp): ScanStats;
+  selectFiles(regex: RegExp, baseDir: string): string[];
+  hash?: string;
+}
+
+class ScanResultsImpl implements ScanResults {
+  /**
+   * Encapsulates the results of a filesystem scan with methods
+   * to easily select scan information or filenames for a
+   * specific subset of the files listed in the scan results.
+   *
+   * @param stats An object that contains filenames
+   *  as keys where each key maps to an object containing the
+   *  hash and number of lines for the specified file.  This
+   *  information is accessed via the `hash` and `lines`
+   *  attributes respectively
+   * @param hash A hashcode computed from the contents of all the files.
+   */
+  constructor(
+      private readonly stats: ScanStats, readonly errorMap: Map<string, Error>,
+      readonly hash?: string) {}
+
+  errors(): Map<string, Error> {
+    return this.errorMap;
+  }
+
+  /**
+   * Used to get all of the file scan results.
+   */
+  all(): ScanStats {
+    return this.stats;
+  }
+
+  /**
+   * Used to get the file scan results for only the files
+   * whose filenames match the specified regex.
+   *
+   * @param {regex} regex The regex that tests a filename
+   *  to determine if the scan results for that filename
+   *  should be included in the returned results.
+   */
+  selectStats(regex: RegExp): ScanStats|{} {
+    return _.pickBy(this.stats, (ignore, key) => {
```

Comment:
If this is the only thing you're bringing in lodash for, you can do this a little less elegantly with a reduce:

```js
Object.keys(object).reduce((acc, key) => {
  if (regex.test(key)) {
    acc[key] = object[key];    
  }
  return acc;
}, {});
```

Reply:
It's also worth asking - if you expect to manage the object this way, should it be a `Map`? (honestly asking, not sure)

Reply:
Aye, I have a PR to remove lodash from that as well.  That's the problem with lodash, you let it in for one method, then it lets in all of it's friends :)

Reply:
This still needs to be addressed.  We changed debug to drop the full lodash dependency and instead pull in `lodash.pickby`.  That or the reduce pattern above should work fine.

Reply:
This is now done.
I got the changes from cloud-profiler-nodejs.